mcp/tools/tools_coder.go

```diff
@@ -0,0 +1,381 @@
+	return func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+		if deps.Client == nil {
+			return nil, xerrors.New("developer error: client is required")
+		}
+		args := request.Params.Arguments
+
+		wsArg, ok := args["workspace"].(string)
+		if !ok {
+			return nil, xerrors.New("workspace is required")
+		}
+
+		workspace, err := getWorkspaceByIDOrOwnerName(ctx, deps.Client, wsArg)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to fetch workspace: %w", err)
+		}
+
+		workspaceJSON, err := json.Marshal(workspace)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to encode workspace: %w", err)
+		}
+
+		return &mcp.CallToolResult{
+			Content: []mcp.Content{
+				mcp.NewTextContent(string(workspaceJSON)),
+			},
+		}, nil
+	}
+}
+
+// Example payload:
+// {"jsonrpc":"2.0","id":1,"method":"tools/call", "params": {"name": "coder_workspace_exec", "arguments": {"workspace": "dev", "command": "ps -ef"}}}
+func handleCoderWorkspaceExec(deps ToolDeps) mcpserver.ToolHandlerFunc {
+	return func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+		if deps.Client == nil {
+			return nil, xerrors.New("developer error: client is required")
+		}
+		args := request.Params.Arguments
+
+		wsArg, ok := args["workspace"].(string)
+		if !ok {
+			return nil, xerrors.New("workspace is required")
+		}
+
+		command, ok := args["command"].(string)
+		if !ok {
+			return nil, xerrors.New("command is required")
+		}
+
+		// Validate the command if allowed commands are specified
+		allowed, err := IsCommandAllowed(command, deps.AllowedExecCommands)
+		if err != nil {
+			return nil, err
+		}
+		if !allowed {
+			return nil, xerrors.Errorf("command not allowed: %s", command)
+		}
+
+		// Attempt to fetch the workspace. We may get a UUID or a name, so try to
+		// handle both.
+		ws, err := getWorkspaceByIDOrOwnerName(ctx, deps.Client, wsArg)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to fetch workspace: %w", err)
+		}
+
+		// Ensure the workspace is started.
+		// Select the first agent of the workspace.
+		var agt *codersdk.WorkspaceAgent
+		for _, r := range ws.LatestBuild.Resources {
+			for _, a := range r.Agents {
+				if a.Status != codersdk.WorkspaceAgentConnected {
+					continue
+				}
+				agt = ptr.Ref(a)
+				break
+			}
+		}
+		if agt == nil {
+			return nil, xerrors.Errorf("no connected agents for workspace %s", ws.ID)
+		}
+
+		conn, err := workspacesdk.New(deps.Client).AgentReconnectingPTY(ctx, workspacesdk.WorkspaceAgentReconnectingPTYOpts{
+			AgentID:     agt.ID,
+			Reconnect:   uuid.New(),
+			Width:       80,
+			Height:      24,
+			Command:     command,
+			BackendType: "buffered", // the screen backend is annoying to use here.
+		})
+		if err != nil {
+			return nil, xerrors.Errorf("failed to open reconnecting PTY: %w", err)
+		}
+		defer conn.Close()
+
+		var buf bytes.Buffer
+		if _, err := io.Copy(&buf, conn); err != nil {
```

Comment:
What if the command is supposed to take more time, should we stream the response?

Reply:
Yeah I'd like to look into this as a future enhancement.